File: apps/sdk-typescript/src/services/observability.ts
/**
 * `ObservabilityService` — typed client for the `observability` service.
 *
 * Mirror of `apps/observability/src/http/routes.ts`:
 *   - POST   /v1/observability/logs
 *   - POST   /v1/observability/traces
 *   - POST   /v1/observability/query
 *   - GET    /v1/observability/alerts
 *   - POST   /v1/observability/alerts
 *   - GET    /v1/observability/alerts/:id
 *   - PATCH  /v1/observability/alerts/:id
 *   - DELETE /v1/observability/alerts/:id   (204 no body)
 *
 * Telemetry ingestion accepts a `events`/`spans` array of unknown shape —
 * the service validates each event imperatively (see
 * `domain/telemetry-validation.ts`) and reports per-index rejections in the
 * response. The SDK doesn't constrain the shape further; callers can pass
 * structured DTOs that match the platform's log/span contract.
 */

import { request } from '../http.js';

import type { HttpClientOptions } from '../http.js';

export type AlertScope = 'logs' | 'spans';

export type FilterOperator =
  | 'eq'
  | 'neq'
  | 'in'
  | 'gt'
  | 'gte'
  | 'lt'
  | 'lte'
  | 'contains'
  | 'matches';

export type FilterValue =
  | string
  | number
  | boolean
  | readonly string[]
  | readonly number[]
  | readonly (string | number)[];

export interface QueryFilter {
  readonly field: string;
  readonly op: FilterOperator;
  readonly value: FilterValue;
}

export interface IngestResponse {
  readonly accepted: number;
  readonly rejected: readonly { readonly index: number; readonly reason: string }[];
}

export interface IngestLogsRequest {
  readonly events: readonly unknown[];
}

export interface IngestSpansRequest {
  readonly spans: readonly unknown[];
}

export interface QueryRequest {
  readonly scope: AlertScope;
  readonly filters?: readonly QueryFilter[];
  readonly timeRange: { readonly from: string; readonly to: string };
  readonly limit?: number;
  readonly offset?: number;
}

export interface QueryRow {
  readonly timestamp: string;
  readonly traceId: string;
  readonly spanId: string;
  readonly parentSpanId?: string;
  readonly service: string;
  readonly agentDid?: string;
  readonly operatorId?: string;
  readonly level?: string;
  readonly message?: string;
  readonly name?: string;
  readonly kind?: string;
  readonly status?: string;
  readonly statusMessage?: string;
  readonly startTimestamp?: string;
  readonly endTimestamp?: string;
  readonly durationMs?: number;
  readonly attributes: Readonly<Record<string, string | number | boolean>>;
  readonly resource: Readonly<Record<string, string>>;
}

export interface QueryResponse {
  readonly rows: readonly QueryRow[];
  readonly total: number;
}

export type AlertCombinator = 'and' | 'or';

export interface AlertCondition {
  readonly operator: AlertCombinator;
  readonly filters: readonly QueryFilter[];
  readonly windowSeconds: number;
  readonly threshold: number;
}

export type NotificationChannel =
  | { readonly type: 'webhook'; readonly url: string }
  | { readonly type: 'slack'; readonly channel: string }
  | { readonly type: 'email'; readonly recipients: readonly string[] };

export interface NotificationConfig {
  readonly channels: readonly NotificationChannel[];
}

export interface AlertCreateRequest {
  readonly ownerOperatorId: string;
  readonly name: string;
  readonly description?: string;
  readonly enabled?: boolean;
  readonly scope: AlertScope;
  readonly condition: AlertCondition;
  readonly cooldownSeconds?: number;
  readonly notification?: NotificationConfig;
}

export interface AlertUpdateRequest {
  readonly name?: string;
  readonly description?: string;
  readonly enabled?: boolean;
  readonly scope?: AlertScope;
  readonly condition?: AlertCondition;
  readonly cooldownSeconds?: number;
  readonly notification?: NotificationConfig;
}

export interface AlertRule {
  readonly id: string;
  readonly ownerOperatorId: string;
  readonly name: string;
  readonly description: string;
  readonly enabled: boolean;
  readonly scope: AlertScope;
  readonly condition: AlertCondition;
  readonly cooldownSeconds: number;
  readonly notification: NotificationConfig;
  readonly createdAt: string;
  readonly updatedAt: string;
}

export interface AlertListResponse {
  readonly alerts: readonly AlertRule[];
}

export class ObservabilityService {
  constructor(
    private readonly opts: HttpClientOptions,
    private readonly baseUrl: string,
  ) {}

  /** POST /v1/observability/logs */
  public async ingestLogs(body: IngestLogsRequest): Promise<IngestResponse> {
    const data = await request<IngestResponse>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/observability/logs',
      body,
    });
    if (data === undefined) {
      throw new Error('observability.ingestLogs: empty response body');
    }
    return data;
  }

  /** POST /v1/observability/traces */
  public async ingestSpans(body: IngestSpansRequest): Promise<IngestResponse> {
    const data = await request<IngestResponse>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/observability/traces',
      body,
    });
    if (data === undefined) {
      throw new Error('observability.ingestSpans: empty response body');
    }
    return data;
  }

  /** POST /v1/observability/query */
  public async query(body: QueryRequest): Promise<QueryResponse> {
    const data = await request<QueryResponse>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/observability/query',
      body,
    });
    if (data === undefined) {
      throw new Error('observability.query: empty response body');
    }
    return data;
  }

  /** GET /v1/observability/alerts?operatorId=... */
  public async listAlerts(operatorId: string): Promise<AlertListResponse> {
    const data = await request<AlertListResponse>(this.opts, {
      method: 'GET',
      baseUrl: this.baseUrl,
      path: '/v1/observability/alerts',
      query: { operatorId },
    });
    if (data === undefined) {
      throw new Error('observability.listAlerts: empty response body');
    }
    return data;
  }

  /** POST /v1/observability/alerts */
  public async createAlert(body: AlertCreateRequest): Promise<AlertRule> {
    const data = await request<AlertRule>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/observability/alerts',
      body,
    });
    if (data === undefined) {
      throw new Error('observability.createAlert: empty response body');
    }
    return data;
  }

  /** GET /v1/observability/alerts/:id */
  public async getAlert(id: string): Promise<AlertRule> {
    const data = await request<AlertRule>(this.opts, {
      method: 'GET',
      baseUrl: this.baseUrl,
      path: `/v1/observability/alerts/${encodeURIComponent(id)}`,
    });
    if (data === undefined) {
      throw new Error('observability.getAlert: empty response body');
    }
    return data;
  }

  /** PATCH /v1/observability/alerts/:id */
  public async patchAlert(id: string, body: AlertUpdateRequest): Promise<AlertRule> {
    const data = await request<AlertRule>(this.opts, {
      method: 'PATCH',
      baseUrl: this.baseUrl,
      path: `/v1/observability/alerts/${encodeURIComponent(id)}`,
      body,
    });
    if (data === undefined) {
      throw new Error('observability.patchAlert: empty response body');
    }
    return data;
  }

  /** DELETE /v1/observability/alerts/:id  (204 No Content). */
  public async deleteAlert(id: string): Promise<void> {
    await request<void>(this.opts, {
      method: 'DELETE',
      baseUrl: this.baseUrl,
      path: `/v1/observability/alerts/${encodeURIComponent(id)}`,
      expectNoBody: true,
    });
  }
}

File: apps/sdk-python/src/colber_sdk/services/observability.py
"""``ObservabilityService`` — typed client for the ``observability`` service.

Mirror of ``apps/sdk-typescript/src/services/observability.ts`` and
``apps/observability/src/http/routes.ts``:

- ``POST   /v1/observability/logs``
- ``POST   /v1/observability/traces``
- ``POST   /v1/observability/query``
- ``GET    /v1/observability/alerts``
- ``POST   /v1/observability/alerts``
- ``GET    /v1/observability/alerts/:id``
- ``PATCH  /v1/observability/alerts/:id``
- ``DELETE /v1/observability/alerts/:id``  (204 No Content)
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal
from urllib.parse import quote

from .._http import HttpClientOptions, RequestParams, request
from ._convert import from_wire, to_wire

AlertScope = Literal["logs", "spans"]


@dataclass(frozen=True, slots=True)
class IngestRejection:
    index: int
    reason: str


@dataclass(frozen=True, slots=True)
class IngestResponse:
    accepted: int
    rejected: list[IngestRejection] = field(default_factory=list)


@dataclass(frozen=True, slots=True)
class QueryRow:
    timestamp: str
    trace_id: str
    span_id: str
    service: str
    attributes: dict[str, Any] = field(default_factory=dict)
    resource: dict[str, str] = field(default_factory=dict)
    parent_span_id: str | None = None
    agent_did: str | None = None
    operator_id: str | None = None
    level: str | None = None
    message: str | None = None
    name: str | None = None
    kind: str | None = None
    status: str | None = None
    status_message: str | None = None
    start_timestamp: str | None = None
    end_timestamp: str | None = None
    duration_ms: float | None = None


@dataclass(frozen=True, slots=True)
class QueryResponse:
    rows: list[QueryRow] = field(default_factory=list)
    total: int = 0


@dataclass(frozen=True, slots=True)
class AlertRule:
    id: str
    owner_operator_id: str
    name: str
    description: str
    enabled: bool
    scope: str
    condition: dict[str, Any]
    cooldown_seconds: int
    notification: dict[str, Any]
    created_at: str
    updated_at: str


@dataclass(frozen=True, slots=True)
class AlertListResponse:
    alerts: list[AlertRule] = field(default_factory=list)


class ObservabilityService:
    """Typed client for the ``observability`` service."""

    def __init__(self, opts: HttpClientOptions, base_url: str) -> None:
        self._opts = opts
        self._base_url = base_url

    def ingest_logs(self, *, events: list[Any]) -> IngestResponse:
        """``POST /v1/observability/logs``."""
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/observability/logs",
                body={"events": events},
            ),
        )
        if data is None:
            raise RuntimeError("observability.ingest_logs: empty response body")
        return from_wire(IngestResponse, data)

    def ingest_spans(self, *, spans: list[Any]) -> IngestResponse:
        """``POST /v1/observability/traces``."""
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/observability/traces",
                body={"spans": spans},
            ),
        )
        if data is None:
            raise RuntimeError("observability.ingest_spans: empty response body")
        return from_wire(IngestResponse, data)

    def query(
        self,
        *,
        scope: AlertScope,
        time_range: dict[str, str],
        filters: list[dict[str, Any]] | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> QueryResponse:
        """``POST /v1/observability/query``."""
        body_in: dict[str, Any] = {"scope": scope, "time_range": time_range}
        if filters is not None:
            body_in["filters"] = filters
        if limit is not None:
            body_in["limit"] = limit
        if offset is not None:
            body_in["offset"] = offset
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/observability/query",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("observability.query: empty response body")
        return from_wire(QueryResponse, data)

    def list_alerts(self, operator_id: str) -> AlertListResponse:
        """``GET /v1/observability/alerts?operatorId=...``."""
        data = request(
            self._opts,
            RequestParams(
                method="GET",
                base_url=self._base_url,
                path="/v1/observability/alerts",
                query={"operatorId": operator_id},
            ),
        )
        if data is None:
            raise RuntimeError("observability.list_alerts: empty response body")
        return from_wire(AlertListResponse, data)

    def create_alert(
        self,
        *,
        owner_operator_id: str,
        name: str,
        scope: AlertScope,
        condition: dict[str, Any],
        description: str | None = None,
        enabled: bool | None = None,
        cooldown_seconds: int | None = None,
        notification: dict[str, Any] | None = None,
    ) -> AlertRule:
        """``POST /v1/observability/alerts``."""
        body_in: dict[str, Any] = {
            "owner_operator_id": owner_operator_id,
            "name": name,
            "scope": scope,
            "condition": condition,
        }
        if description is not None:
            body_in["description"] = description
        if enabled is not None:
            body_in["enabled"] = enabled
        if cooldown_seconds is not None:
            body_in["cooldown_seconds"] = cooldown_seconds
        if notification is not None:
            body_in["notification"] = notification
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/observability/alerts",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("observability.create_alert: empty response body")
        return from_wire(AlertRule, data)

    def get_alert(self, alert_id: str) -> AlertRule:
        """``GET /v1/observability/alerts/:id``."""
        data = request(
            self._opts,
            RequestParams(
                method="GET",
                base_url=self._base_url,
                path=f"/v1/observability/alerts/{quote(alert_id, safe='')}",
            ),
        )
        if data is None:
            raise RuntimeError("observability.get_alert: empty response body")
        return from_wire(AlertRule, data)

    def patch_alert(
        self,
        alert_id: str,
        *,
        name: str | None = None,
        description: str | None = None,
        enabled: bool | None = None,
        scope: AlertScope | None = None,
        condition: dict[str, Any] | None = None,
        cooldown_seconds: int | None = None,
        notification: dict[str, Any] | None = None,
    ) -> AlertRule:
        """``PATCH /v1/observability/alerts/:id``."""
        body_in: dict[str, Any] = {}
        if name is not None:
            body_in["name"] = name
        if description is not None:
            body_in["description"] = description
        if enabled is not None:
            body_in["enabled"] = enabled
        if scope is not None:
            body_in["scope"] = scope
        if condition is not None:
            body_in["condition"] = condition
        if cooldown_seconds is not None:
            body_in["cooldown_seconds"] = cooldown_seconds
        if notification is not None:
            body_in["notification"] = notification
        data = request(
            self._opts,
            RequestParams(
                method="PATCH",
                base_url=self._base_url,
                path=f"/v1/observability/alerts/{quote(alert_id, safe='')}",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("observability.patch_alert: empty response body")
        return from_wire(AlertRule, data)

    def delete_alert(self, alert_id: str) -> None:
        """``DELETE /v1/observability/alerts/:id``  (204 No Content)."""
        request(
            self._opts,
            RequestParams(
                method="DELETE",
                base_url=self._base_url,
                path=f"/v1/observability/alerts/{quote(alert_id, safe='')}",
                expect_no_body=True,
            ),
        )
